plugin.js

```diff
@@ -0,0 +1,112 @@
+'use strict';
+
+const PLUGIN_ID_KEY = '___shelljs_plugin_id___';
+const DEFAULT_ERROR_CODE = 1;
+
+function setPluginID(plugin, id) {
+  Object.defineProperty(plugin, PLUGIN_ID_KEY, {
+    value: id
+  });
+  return id;
+}
+
+const getPluginID = plugin => plugin[PLUGIN_ID_KEY];
+
+module.exports = function setupShellPlugins($) {
+  const used_plugins = [];
+  const plugins = [{ // $
```

Comment:
Plugin ID #0 is always the initial shelljs object, and as such has to be created and setup here, with some values hardcoded.